File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/ModalPlacement.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.property;

import com.urbanairship.android.layout.model.SafeAreaAware;
import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import static com.urbanairship.android.layout.model.SafeAreaAware.ignoreSafeAreaFromJson;

public class ModalPlacement implements SafeAreaAware {

    @NonNull
    private final ConstrainedSize size;
    @Nullable
    private final Margin margin;
    @Nullable
    private final Position position;
    @Nullable
    private final Color shadeColor;
    private final boolean ignoreSafeArea;

    @Nullable
    private final Orientation orientationLock;

    @Nullable
    private final Border border;

    @Nullable
    private final Color backgroundColor;

    public ModalPlacement(
            @NonNull ConstrainedSize size,
            @Nullable Margin margin,
            @Nullable Position position,
            @Nullable Color shadeColor,
            boolean ignoreSafeArea,
            @Nullable Orientation orientationLock,
            @Nullable Border border,
            @Nullable Color backgroundColor
    ) {
        this.size = size;
        this.margin = margin;
        this.position = position;
        this.shadeColor = shadeColor;
        this.ignoreSafeArea = ignoreSafeArea;
        this.orientationLock = orientationLock;
        this.border = border;
        this.backgroundColor = backgroundColor;
    }

    @NonNull
    public static ModalPlacement fromJson(@NonNull JsonMap json) throws JsonException {
        JsonMap sizeJson = json.opt("size").optMap();
        if (sizeJson.isEmpty()) {
            throw new JsonException("Failed to parse Modal Placement! Field 'size' is required.");
        }
        JsonMap positionJson = json.opt("position").optMap();
        JsonMap marginJson = json.opt("margin").optMap();
        JsonMap borderJson = json.opt("border").optMap();
        JsonMap backgroundJson = json.opt("background_color").optMap();

        ConstrainedSize size = ConstrainedSize.fromJson(sizeJson);
        Margin margin = marginJson.isEmpty() ? null : Margin.fromJson(marginJson);
        Position position = positionJson.isEmpty() ? null : Position.fromJson(positionJson);
        Color shadeColor = Color.fromJsonField(json, "shade_color");
        boolean ignoreSafeArea = ignoreSafeAreaFromJson(json);

        String orientationString = json.opt("device").optMap().opt("lock_orientation").optString();
        Orientation orientationLock = orientationString.isEmpty() ? null : Orientation.from(orientationString);

        Border border = borderJson.isEmpty() ? null : Border.fromJson(borderJson);
        Color backgroundColor = backgroundJson.isEmpty() ? null : Color.fromJson(backgroundJson);

        return new ModalPlacement(size, margin, position, shadeColor, ignoreSafeArea, orientationLock, border, backgroundColor);
    }

    @Nullable
    public Orientation getOrientationLock() {
        return orientationLock;
    }

    @Nullable
    public Margin getMargin() {
        return margin;
    }

    @NonNull
    public ConstrainedSize getSize() {
        return size;
    }

    @Nullable
    public Position getPosition() {
        return position;
    }

    @Nullable
    public Color getShadeColor() {
        return shadeColor;
    }

    @Override
    public boolean shouldIgnoreSafeArea() {
        return ignoreSafeArea;
    }

    @Nullable
    public Border getBorder() {
        return border;
    }

    @Nullable
    public Color getBackgroundColor() {
        return backgroundColor;
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/property/ModalPlacement.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.property

import androidx.annotation.RestrictTo
import com.urbanairship.android.layout.model.SafeAreaAware
import com.urbanairship.android.layout.model.SafeAreaAware.Companion.ignoreSafeAreaFromJson
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonMap
import com.urbanairship.json.JsonValue
import com.urbanairship.json.optionalField
import com.urbanairship.json.optionalList
import com.urbanairship.json.optionalMap
import com.urbanairship.json.requireField

/** @hide */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class ModalPlacement(
    public val size: ConstrainedSize,
    public val margin: Margin?,
    public val position: Position?,
    public val shadeColor: Color?,
    private val ignoreSafeArea: Boolean,
    public val border: Border?,
    public val backgroundColor: Color?,
    public val shadow: Shadow?,
    public val animation: ModalAnimation?
) : SafeAreaAware {

    public override fun shouldIgnoreSafeArea(): Boolean {
        return ignoreSafeArea
    }

    public companion object {
        private const val KEY_SIZE = "size"
        private const val KEY_POSITION = "position"
        private const val KEY_MARGIN = "margin"
        private const val KEY_BORDER = "border"
        private const val KEY_BACKGROUND_COLOR = "background_color"
        private const val KEY_SHADE_COLOR = "shade_color"
        private const val KEY_SHADOW = "shadow"
        private const val KEY_SELECTORS = "selectors"
        private const val KEY_ANIMATION = "animation"

        @Throws(JsonException::class)
        public fun fromJson(json: JsonValue): ModalPlacement {
            val content = json.requireMap()

            val shadow = content
                .optionalMap(KEY_SHADOW)
                ?.optionalList(KEY_SELECTORS)
                ?.map(ShadowSelector::fromJson)
                ?.firstOrNull { it.platform == null || it.platform == Platform.ANDROID }
                ?.shadow

            return ModalPlacement(
                size = ConstrainedSize.fromJson(content.require(KEY_SIZE)),
                margin = content[KEY_MARGIN]?.let(Margin::fromJson),
                position = content[KEY_POSITION]?.let(Position::fromJson),
                shadeColor = content[KEY_SHADE_COLOR]?.let(Color::fromJson),
                ignoreSafeArea = ignoreSafeAreaFromJson(content),
                border = content[KEY_BORDER]?.let(Border::fromJson),
                backgroundColor = content[KEY_BACKGROUND_COLOR]?.let(Color::fromJson),
                shadow = shadow,
                animation = content[KEY_ANIMATION]?.let(ModalAnimation::fromJson)
            )
        }
    }
}
